### DISCUSSION TITLE
Best Practice on Machine-to-Machine Authentication?

### DISCUSSION THREAD

**Discussion by forsberg** [keycloak/keycloak#32638] [Q&A]
Hi!

We're using Keycloak as IdP for an API, with our customers as users. Currently, there's a web portal and several mobile apps talking to the API. Our customers are logging in using their usernames and passwords, and everything is working great, keycloak really has made life better for us compared to our old homegrown solution.

However, we also see a need for our customers to be able to setup various small devices that need to talk to the API to retrieve data. One example could be some kind of display that visualize some of the data from our products. 

Obviously, it would be bad practice to use personal accounts and credentials for this. Instead, it would make sense to provide a UI for the human users where they could extract some kind of token for the device to use when authenticating to our API, where the token would have limited permissions, i.e limited scopes and/or other limitations. A bit like [Project Access Tokens in Gitlab](https://docs.gitlab.com/ee/user/project/settings/project_access_tokens.html).

I guess I could create OIDC clients for each device, but as we're talking what could potentially be thousands of devices, I'm not sure this is the best option.

What is best practice in the Keycloak community to solve this kind of problem? Any authentication standards we should use? My Google and LLM skills are failing me on this one.

**Comment by stianst** [comment]
If the device should act on behalf of itself use the client credential grant and a service account. If the device should act on behalf of a user use an offline token (requested from an app using scope=offline)/
